volatility3/framework/plugins/windows/indirect_system_calls.py

```diff
@@ -0,0 +1,124 @@
+        super().__init__(*args, **kwargs)
+
+        self.syscall_finder = direct_system_calls.syscall_finder_type(
+            # gets the target address of a indirect jmp
+            self._indirect_syscall_block_target,
+            # we are looking for indirect system calls, so we don't want 'syscall' instructions in our code block
+            False,
+            # jmp [address]; ret
+            "/\\xff\\x25[^\\xc3]{,24}\\xc3/",
+            # any of these mean we aren't in a malicious indirect call
+            ["call", "leave", "int3", "ret"],
+            # stop at jmp, this should reference the system call instruction
+            ["jmp"],
+        )
+
+    @classmethod
+    def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]:
+        # create a list of requirements for vadyarascan
+        vadyarascan_requirements = [
+            requirements.ModuleRequirement(
+                name="kernel",
+                description="Windows kernel",
+                architectures=["Intel32", "Intel64"],
+            ),
+            requirements.PluginRequirement(
+                name="pslist", plugin=pslist.PsList, version=(2, 0, 0)
+            ),
+            requirements.VersionRequirement(
+                name="yarascanner", component=yarascan.YaraScanner, version=(2, 1, 0)
+            ),
+            requirements.PluginRequirement(
+                name="yarascan", plugin=yarascan.YaraScan, version=(2, 0, 0)
+            ),
+            requirements.PluginRequirement(
+                name="direct_system_calls",
+                plugin=direct_system_calls.DirectSystemCalls,
+                version=(1, 0, 0),
+            ),
+        ]
+
+        # get base yarascan requirements for command line options
+        yarascan_requirements = yarascan.YaraScan.get_yarascan_option_requirements()
+
+        # return the combined requirements
+        return yarascan_requirements + vadyarascan_requirements
+
+    @staticmethod
+    def _indirect_syscall_block_target(
+        proc_layer: interfaces.layers.DataLayerInterface, inst: capstone._cs_insn
+    ) -> Optional[int]:
+        """
+        This function determines the address of a jmp in the following form:
+
+        jmp [address]
+
+        To determine this, we must:
+        1) Pull the 4 byte relative offset of 'address' inside the instruction
+        2) Compute the full address of this relative offset
+        3) Read from the address as it is being dereferenced
+        4) Ensure the target address points to a 'syscall' instruction
+
+        Args:
+            proc_layer: the layer of the potential syscall block
+            inst: the terminating instruction of the syscall block check
+        Returns:
+            The target address of the jump if it can be computed
+        """
+
+        try:
+            jmp_address_str = proc_layer.read(inst.address, 6)
+        except exceptions.InvalidAddressException:
+            return None
```

Comment:
I assume we'd hit these so often that even adding debugging messages here to say which outcome was hit, wouldn't be worthwhile.  I just worry there's lots of ways this method can terminate, and you might want to know whether it was the very last one rather than the very first?  I know the debug log spews out a lot already, so I'm not gonna push it, just a question as to whether it would be more help than hinderance?

Reply:
It wouldn't help. It would essentially be telling people that a `\xff\x25` byte pair we found in memory wasn't actually a jump. Others would be telling that the integer after a `\xff\x25` don't correspond to a valid offset. Across all processes you would likely have tens of 1000s of these being reported.